multi/mcxn94x-multi/cpu.c

```diff
@@ -0,0 +1,275 @@
+				}
+				else {
+					break;
+				}
+			}
+		}
+	}
+	else {
+		if (lf_fifo_empty(&common.rx_fifo) != 0) {
+			mutexUnlock(common.rx_lock);
+			return -EWOULDBLOCK;
+		}
+
+		while (bytes_read < size) {
+			if (lf_fifo_pop(&common.rx_fifo, &buf[bytes_read]) != 0) {
+				bytes_read += 1;
+			}
+			else {
+				break;
+			}
+		}
+	}
+
+	mutexUnlock(common.rx_lock);
+
+	return (ssize_t)bytes_read;
+}
+
+
+static ssize_t fifoWrite(const unsigned char *buf, size_t len, unsigned int blocking)
+{
+	size_t to_write;
+	size_t bytes_written = 0;
+
+	if (buf == NULL || len == 0) {
+		return -EINVAL;
+	}
+
+	/* Synchronize multiple writers */
+	mutexLock(common.tx_lock);
+
+	while (bytes_written < len) {
+		if (blocking == 0 && REG_IRQ(common.coreid_remote, reg_irq) != 0u) {
+			mutexUnlock(common.tx_lock);
+			return (bytes_written > 0) ? bytes_written : -EAGAIN;
+		}
+
+		while (blocking != 0 && REG_IRQ(common.coreid_remote, reg_irq) != 0u) {
+			; /* Wait until the other CPU reads IRQx register */
+		}
+
+		/* Message structure: MSB[ 6b reserved | 2b len(data) | 24b data ]LSB */
+		uint32_t value = 0;
+		to_write = min(len - (size_t)bytes_written, 3);
+
+		for (size_t i = 0; i < to_write; i++) {
+			value |= (uint32_t)buf[bytes_written] << (8 * i);
+			bytes_written += 1;
+		}
+
+		value |= ((uint32_t)to_write & 0x3) << 24;
+
+		REG_IRQ(common.coreid_remote, reg_irq_set) = value;
+	}
+
+	mutexUnlock(common.tx_lock);
+
+	return (ssize_t)bytes_written;
+}
+
+
+static int fifoPollStatus(void)
+{
+	int revents = 0;
+
+	if (lf_fifo_empty(&common.rx_fifo) == 0) {
+		revents |= POLLIN | POLLRDNORM;
+	}
+
+	if (REG_IRQ(common.coreid_remote, reg_irq) == 0u) {
+		revents |= POLLOUT | POLLWRNORM;
+	}
+
+	return revents;
+}
+
+
+static void cpu_handleMsg(msg_t *msg, msg_rid_t rid, unsigned int major, unsigned int minor)
+{
+	unsigned int blocking;
+
+	if (minor > 0) {
+		msg->o.err = -EINVAL;
+		msgRespond(msg->oid.port, msg, rid);
+		return;
+	}
+
+	switch (msg->type) {
+		case mtOpen:
+		case mtClose:
+			msg->o.err = 0;
+			break;
+
+		case mtWrite:
+			blocking = ((msg->i.io.mode & O_NONBLOCK) != 0u) ? 0u : 1u;
+			msg->o.err = fifoWrite(msg->i.data, msg->i.size, blocking);
+			break;
+
+		case mtRead:
+			blocking = ((msg->i.io.mode & O_NONBLOCK) != 0u) ? 0u : 1u;
+			msg->o.err = fifoRead(msg->o.data, msg->o.size, blocking);
+			break;
+
+		case mtGetAttr:
+			if (msg->i.attr.type != atPollStatus) {
+				msg->o.err = -ENOSYS;
+				break;
+			}
+			msg->o.attr.val = fifoPollStatus();
+			msg->o.err = 0;
+			break;
+
+		default:
+			msg->o.err = -ENOSYS;
+			break;
+	}
+
+	msgRespond(msg->oid.port, msg, rid);
+}
+
+
+static void cpu_init(void)
+{
+	common.mailbox = (void *)0x400b2000;
+
+	lf_fifo_init(&common.rx_fifo, common.rx_buf, sizeof(common.rx_buf));
+
+	mutexCreate(&common.rx_lock);
+	mutexCreate(&common.tx_lock);
+	mutexCreate(&common.irq_lock);
+	condCreate(&common.rx_cond);
+
+	interrupt(mailbox0_irq, mailbox_irqHandler, NULL, common.rx_cond, NULL);
```

Comment:
In interrupt we're using `common.coreid_self`. It's dangerous to enable the interrupt before this value is set

Reply:
Done